x-pack/legacy/plugins/maps/public/layers/sources/source_registry.ts

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+/* eslint-disable @typescript-eslint/consistent-type-definitions */
+
+import { AbstractSourceDescriptor } from '../../../common/descriptor_types';
+import { ISource } from './source';
+
+type SourceRegistryEntry = {
+  factory: (sourceDescriptor: AbstractSourceDescriptor, inspectorAdapters: object) => ISource;
+  type: string;
+};
+
+const registry: SourceRegistryEntry[] = [];
+
+export function registerSource(entry: SourceRegistryEntry) {
```

Comment:
needs to check if `entry.type` is not duplicated or throw. Unlike for the layer-wizard, these types persist in the state-store and need to be instantiated in POJO.